How to deploy (#5177)
RasaHQ/rasa: docs/user-guide/how-to-deploy.rst

> :desc: How to deploy your Rasa Assistant with Docker or Kubernetes

Reviewer: ```suggestion
:desc: How to deploy your Rasa Assistant with Docker Compose or Kubernetes
```

.. _deploying-your-rasa-assistant:

Deploying your Rasa Assistant
===============================

.. edit-link::

This page explains when and how to deploy an assistant built with Rasa. 
It will allow you to make your assistant available to users and set you up with a production-ready environment.

> If you haven't used Rasa before, we'd recommend that you start with the :ref:`rasa-tutorial`.

Reviewer: is this sentence necessary? I feel like people navigating to this page will have started with rasa already

Author: sure, something similar is on the tutorial pages but people would be going there first so it's not necessary here

Author: removed

.. contents::
   :local:

When to deploy your assistant
--------------------------------

> The best time to deploy your assistant and make it available to test users is once it can handle the most important happy paths, or is what we call a "minimum viable assistant".

Reviewer 2: ```suggestion
The best time to deploy your assistant and make it available to test users is once you have built a :ref:`minimum viable assistant <glossary>`, one that can handle the most important happy paths.
```

Reviewer 2: I am not sure that is quite the right syntax, but link to the glossary

Author: I can link to the glossary page itself, but couldn't find a way to reference a specific item in the glossary  since it is not a section header.

Author: adding a reference in the glossary `.. _minimum-viable-assistant:` then referring to it with `minimum-viable-assistant` or `minimum viable assistant <minimum-viable-assistant>` didn't quite work

Reviewer 2: Yeah, that is fine. I think I ran into the same issue

Author: linked

> The recommended deployment methods described below make it easy to share your assistant with test users via the `share your bot feature in Rasa X <../../rasa-x/docs/user-guide/enable-workflows#conversations-with-test-users>`_, even before selecting which channel you want to use. Then, when you’re ready to make your assistant available via :ref:`messaging-and-voice-channels`, you can easily add them to your existing deployment set up.

Reviewer: ```suggestion
The recommended deployment methods described below make it easy to share your assistant with test users via the `share your bot feature in Rasa X <../../rasa-x/docs/user-guide/enable-workflows#conversations-with-test-users>`_, even before deciding to put it live to your end users. Then, when you’re ready to make your assistant available via :ref:`messaging-and-voice-channels`, you can easily add them to your existing deployment set up.
```
"selecting which channel you want to use" sounds a bit confusing. 
Also is the feature called "share your bot" or "share your assistant"?

Author: changed to 'share your assistant' ; references elsewhere say that

> The recommended deployment methods described below make it easy to share your assistant with test users via the `share your bot feature in Rasa X <../../rasa-x/docs/user-guide/enable-workflows#conversations-with-test-users>`_, even before selecting which channel you want to use. Then, when you’re ready to make your assistant available via one or more :ref:`messaging-and-voice-channels`, you can easily add them to your existing deployment set up.

Reviewer 2: ```suggestion
The recommended deployment methods described below make it easy to share your assistant with test users via the `share your bot feature in Rasa X <../../rasa-x/docs/user-guide/enable-workflows#conversations-with-test-users>`_, even before selecting which channel you want to use. Then, when you’re ready to make your assistant available via :ref:`messaging-and-voice-channels`, you can easily add them to your existing deployment set up.
```

.. _recommended-deployment-methods:

Recommended Deployment Methods
------------------------------

> The recommended way to deploy an assistant is using either the Docker-Compose or Kubernetes options we support. Both deploy Rasa X and your assistant. They are the easiest ways to deploy your assistant, allow you to use Rasa X to view conversations and turn them into training data, and are production-ready.

Reviewer: ```suggestion
The recommended way to deploy an assistant is using either the Docker Compose or Kubernetes options we support. Both deploy Rasa X and your assistant. They are the easiest ways to deploy your assistant, allow you to use Rasa X to view conversations and turn them into training data, and are production-ready.
```

> Kubernetes

Reviewer: Kubernetes/Openshift

Author: added

~~~~~~~~~~

> Kubernetes is the best option if you plan to serve many users. It's straightforward to deploy if you use the helm charts we provide. However, you can also customize the Helm charts if you have specific requirements.

Reviewer: should we maybe say "need a scalable architecture"? Rather than plan to server many users? because many users is also quite arbitrary

Author: changed

    - Default: Read the docs `here <../../rasa-x/docs/installation-and-setup/openshift-kubernetes/>`_.
    - Custom: Read the docs `here <../../rasa-x/docs/installation-and-setup/openshift-kubernetes/>`_ and customize the `open source Helm charts <https://github.com/RasaHQ/rasa-x-helm>`_.

> Docker-Compose

Reviewer: ```suggestion
Docker Compose
```

~~~~~~~~~~~~~~

>     - Default: Watching this video or read the docs `here <../../rasa-x/docs/installation-and-setup/docker-compose-script/>`_.

Reviewer: which video?

Author: link added

    - Custom: Read the docs `here <../../rasa-x/docs/installation-and-setup/docker-compose-manual/>`_.

.. _rasa-only-deployment:

Rasa-only Deployment
----------------------

> Although it is not the recommended deployment method, it is also possible to deploy a Rasa assistant using Docker without Rasa X. See :ref:`running-rasa-with-docker` for details. 

Reviewer: as ty/alan mentioned, bit too intense and we should move running rasa with docker here :)

Author: moved in

> Although it is not the recommended deployment method, it is also possible to deploy a Rasa assistant using Docker without Rasa X. See :ref:`running-rasa-with-docker` for details. 

Reviewer 2: I think we should move the entire running rasa with docker section below `Rasa-only Deployment`

Reviewer 3: also 'not the recommended deployment' is too aggressive IMO.

Author: sure, will move it in and change the wording.

Reviewer 2: That is true. By saying the other route is the recommended deployment, we implicitly are saying this anyway. We can just cut out everything before "It is also possible to deploy..."

:desc: How to deploy your Rasa Assistant with Docker Compose or Kubernetes/Openshift

.. _deploying-your-rasa-assistant:

Deploying your Rasa Assistant

> ===============================

Author: Will change - I have been viewing with `make livedocs`, but I probably haven't been looking out for the right details

> ===============================

Author: fixed lines

> ===============================

Reviewer: there should have been warnings/it shouldn't be visualised correctly - e.g. with the `here` stuff it threw warnings for me

.. edit-link::

This page explains when and how to deploy an assistant built with Rasa. 
It will allow you to make your assistant available to users and set you up with a production-ready environment.

.. contents::
   :local:
   :depth: 2

When to deploy your assistant
--------------------------------

.. raw:: html

>     The best time to deploy your assistant and make it available to test users is once it can handle the most important happy paths, or is what we call a <a style="text-decoration: none" href="https://rasa.com/docs/rasa/glossary">minimum viable assistant</a>.

Reviewer 2: [](url)
```suggestion
    The best time to deploy your assistant and make it available to test users is once it can handle the most important happy paths or is what we call a <a style="text-decoration: none" href="https://rasa.com/docs/rasa/glossary">minimum viable assistant</a>.
```

The recommended deployment methods described below make it easy to share your assistant with test users via the `share your assistant feature in Rasa X <../../rasa-x/docs/user-guide/enable-workflows#conversations-with-test-users>`_. Then, when you’re ready to make your assistant available via one or more :ref:`messaging-and-voice-channels`, you can easily add them to your existing deployment set up.

.. _recommended-deployment-methods:

Recommended Deployment Methods
------------------------------

The recommended way to deploy an assistant is using either the Docker Compose or Kubernetes/Openshift options we support. Both deploy Rasa X and your assistant. They are the easiest ways to deploy your assistant, allow you to use Rasa X to view conversations and turn them into training data, and are production-ready.

Kubernetes/Openshift
~~~~~~~~~~~~~~~~~~~~

> Kubernetes/Openshift is the best option if you plan to serve many users. It's straightforward to deploy if you use the helm charts we provide. However, you can also customize the Helm charts if you have specific requirements.

Reviewer: ```suggestion
Kubernetes/Openshift is the best option if you need a scalable architecture. It's straightforward to deploy if you use the helm charts we provide. However, you can also customize the Helm charts if you have specific requirements.
```

>     - Default: Read the docs `here <../../rasa-x/docs/installation-and-setup/openshift-kubernetes/>`_.

Reviewer: you also need to add two underscores at the end of the link, otherwise the targets won't render properly

Reviewer: for each "here" reference: https://github.com/sphinx-doc/sphinx/issues/3921#issuecomment-315581557

Author: added
